File: examples/dangling_pointer_heap.c
#include "stdlib.h"
#include "stdio.h"

#define BUF_SIZE 10

int main() {
    int *array = malloc(BUF_SIZE * sizeof(int));
    for (size_t index = 0; index < BUF_SIZE; index++) {
        array[index] = index;
    }
    // Take a pointer into the heap-allocated array
    int *ptr = &array[6];

    // at this point, ptr becomes a dangling pointer
    free(array);

    // Even clang does not seem to detect this yet
    printf("%d's array has been set free!\n", *ptr);
    *ptr = 3;
    printf("array[6] was modified to %d after being free'd!\n", *ptr);
}

File: examples/dangling_pointer_heap.rs
fn main() {
    /* Vec is heap-allocated, growing (reallocating) array */
    let mut array = Vec::new();
    for index in 0..10 {
        array.push(index as u32);
    }

    // Take a pointer into the heap-allocated array
    let reference = &mut array[6];

    // at this point, reference becomes a dangling pointer
    drop(array);

    println!("{}'s array has been set free!", *reference);
    *reference = 3;
    println!("array[6] was modified to {} after being free'd!", *reference);
}
